### PR TITLE
pgo: profile-guided optimization for golang v1.17

### PR DESCRIPTION
**Pull request by rajbarik**
This PR implements profile-guided optimization for Golang v1.17. At a very high-level, our implementation consumes PPROF CPU profiles via the compiler flag '-fpprofprofileuse' and performs three key optimizations based on profiles: code specialization ('-specialize'), inlining ('-pprofinline'), code layout ('-codelayout'). Code specialization targets interface method calls to specialize them to direct function calls. Inlining decisions are made based on profiles. Codelayout optimization currently reorders functions based on profiles. BasicBlock layout optimization is currently WIP.

Inlining based on profiles yields decent wins (avg 15%, max 40%). 

For [bitset](https://github.com/bits-and-blooms/bitset), we see ~19% gain from inlining. Please find the steps below to reproduce:

_(bitset)  go test -bench=LemireIterateManyb -cpu 1 -parallel 1 -benchtime=3x -cpuprofile LemireIterate.pprof
pkg: github.com/bits-and-blooms/bitset
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkLemireIterateManyb 	       3	1780585980 ns/op
PASS
ok  	github.com/bits-and-blooms/bitset	12.828s_

_(bitset)   go test -bench=LemireIterateManyb -cpu 1 -parallel 1 -benchtime=3x -gcflags="-fpprofprofileuse LemireIterate.pprof -specialize -pprofinline"
goos: darwin
goarch: amd64
pkg: github.com/bits-and-blooms/bitset
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkLemireIterateManyb 	       3	1597644922 ns/op
PASS
ok  	github.com/bits-and-blooms/bitset	10.303s_

Please note: due to Go compiler dependency issues, we had to duplicate a lot of functionalities from PPROF and from inline/inl.go in order to make this work. We are currently working closely with Google-go folks to eliminate these redundancies.

### CODE REVIEW

**Comment by mvdan** [comment]
Did you see https://github.com/golang/go/issues/28262? Usually, big features like these are designed and discussed in the issue tracker first. This is a very large amount of code to show up with all of a sudden :)

**Comment by ianlancetaylor** [comment]
@mvdan and others: for background, the authors of this CL have been discussing their PGO implementation privately with the Go compiler team.  This pull request is the mechanism they are using to share their code changes.  This pull request certainly isn't going to go in as is--for one thing, it's against the 1.17 branch--but will be used for further discussion and, I hope, a later updated implementation on mainline.

So, it's new to the community (and new to me too, for that matter) but it's not entirely out of the blue.

**Comment by mvdan** [comment]
Thanks for the quick reply, @ianlancetaylor! I had some thoughts but they weren't related to the code changes, so I've left a comment on the issue instead.

**Comment by rajbarik** [comment]
Thank you for proving the context @ianlancetaylor. This PR is still WIP as we (Uber) are working closely with @cherrymui @aclements and @jeremyfaller on cleaning up compiler dependencies (which would avoid a large amount of code duplication), supporting the common profiling format, adding bblayout optimization, and integrating our code in the main branch.

**Comment by josharian** [comment]
Usually changes of this magnitude start with a publicly circulated design proposal.

**Comment by jeremyfaller** [comment]
> Usually changes of this magnitude start with a publicly circulated design proposal.

This will be no different. I don't think anyone is implying we would skip the proposal process. Uber's code drop is to get their work and ours on the same page, so we can build from there. I don't know what form it will take, but we have to start by sharing. :)

**Comment by prattmic** [comment]
Minor note: @google-cla thinks there is no CLA signed yet. Per https://opensource.google/docs/cla/#external-contributors, I believe you need to get added to the appropriate "Google CLA" Google Group within Uber to be seen as from Uber by the bot.

**Comment by lni** [comment]
@rajbarik, thanks for sharing this very interesting WIP patch! 

When building your feature branch at https://github.com/rajbarik/go/tree/pgo, I got the following errors 

```
~/src/go/src$ ./make.bash
Building Go cmd/dist using /usr/local/go. (go1.17.2 darwin/arm64)
Building Go toolchain1 using /usr/local/go.
# bootstrap/cmd/compile/internal/pgo
/Users/lni/src/go/src/cmd/compile/internal/pgo/irgraph.go:230: fn.NodeFreq undefined (type *ir.Func has no field or method NodeFreq)
/Users/lni/src/go/src/cmd/compile/internal/pgo/irgraph.go:230: undefined: ir.NodeFrequency
/Users/lni/src/go/src/cmd/compile/internal/pgo/irgraph.go:258: call.EdgeFreq undefined (type *ir.CallExpr has no field or method EdgeFreq)
/Users/lni/src/go/src/cmd/compile/internal/pgo/irgraph.go:259: call.EdgeFreq undefined (type *ir.CallExpr has no field or method EdgeFreq)
/Users/lni/src/go/src/cmd/compile/internal/pgo/irgraph.go:259: undefined: ir.EdgeFrequency
/Users/lni/src/go/src/cmd/compile/internal/pgo/irgraph.go:262: call.EdgeFreq undefined (type *ir.CallExpr has no field or method EdgeFreq)
/Users/lni/src/go/src/cmd/compile/internal/pgo/irgraph.go:262: undefined: ir.EdgeFrequency
/Users/lni/src/go/src/cmd/compile/internal/pgo/irgraph.go:303: f.NodeFreq undefined (type *ir.Func has no field or method NodeFreq)
/Users/lni/src/go/src/cmd/compile/internal/pgo/irgraph.go:303: undefined: ir.NodeFrequency
go tool dist: FAILED: /usr/local/go/bin/go install -gcflags=-l -tags=math_big_pure_go compiler_bootstrap bootstrap/cmd/...: exit status 2
``` 

quick grep search shows ir.NodeFrequency is indeed missing from the ir package. 

Did I miss something here? Thanks.

**Comment by rajbarik** [comment]
@lni My apologies. I have updated the PR now and have tested it to work.

@cherrymui fyi.

**Comment by rajbarik** [comment]
> Minor note: @google-cla thinks there is no CLA signed yet. Per https://opensource.google/docs/cla/#external-contributors, I believe you need to get added to the appropriate "Google CLA" Google Group within Uber to be seen as from Uber by the bot. This will be needed eventually before changes can go in, though as noted above, design docs / proposals will come well before that.

Thanks for pointing this out. I am working on getting this fixed.

**Comment by lni** [comment]
@rajbarik thanks a lot for the quick response. 

I've managed to build your patched version of go, interestingly I couldn't reproduce the above reported performance improvements. That LemireIterateManyb benchmark from bitset is slower when PGO is enabled. Please see my results below. 

```
$ /Users/lni/src/go/bin/go test -bench=LemireIterateManyb -cpu 1 -parallel 1 -benchtime=3x -cpuprofile LemireIterate.pprof
goos: darwin
goarch: amd64
pkg: github.com/bits-and-blooms/bitset
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
BenchmarkLemireIterateManyb 	       3	   2676432 ns/op
PASS
ok  	github.com/bits-and-blooms/bitset	0.230s

$ /Users/lni/src/go/bin/go test -bench=LemireIterateManyb -cpu 1 -parallel 1 -benchtime=3x -gcflags="-fpprofprofileuse LemireIterate.pprof -specialize -pprofinline"
goos: darwin
goarch: amd64
pkg: github.com/bits-and-blooms/bitset
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
BenchmarkLemireIterateManyb 	       3	   2775468 ns/op
PASS
ok  	github.com/bits-and-blooms/bitset	0.233s
```

two observed differences in our benchmark outputs, 

1. your ns/op is significantly higher than mine, 1780585980 ns/op vs. 2676432 ns/op
2. different processors, i7-8850H processor at 2.6Ghz vs. i9-9980HK at 2.4Ghz 

I've tried on a M1 ARM processor based mbp as well and I got similar results. 

Did I still miss anything here? Thanks!

**Comment by rajbarik** [comment]
@lni Great to hear that you are able to build and test the patch. 
AFAICT, performance improvements may vary between hardware generations and SKUs. You may want to increase the benchtime to 10x or more to increase the execution time [in order to eliminate any noises]. Also, I recommend adding "-m" to gcflags and mining the logs to ensure inlining via pprof profiles is indeed working as expected [before and after].  Please see example below:

**Before pprofinline**

(bitset) go test -bench=LemireIterateManyb -cpu 1 -parallel 1 -benchtime=3x -gcflags="-m" -cpuprofile LemireIterate.pprof 2> before_log.txt
goos: darwin
goarch: amd64
pkg: github.com/bits-and-blooms/bitset
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkLemireIterateManyb 	       3	1483586757 ns/op
PASS
ok  	github.com/bits-and-blooms/bitset	11.476s

(bitset) go tool pprof LemireIterate.pprof
Type: cpu
Time: Dec 6, 2021 at 9:42am (PST)
Duration: 10.74s, Total samples = 9.29s (86.47%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top20
Showing nodes accounting for 9.21s, 99.14% of 9.29s total
Dropped 19 nodes (cum <= 0.05s)
      flat  flat%   sum%        cum   cum%
     3.29s 35.41% 35.41%      3.39s 36.49%  github.com/bits-and-blooms/bitset.(*BitSet).NextSetMany
     2.69s 28.96% 64.37%      9.11s 98.06%  github.com/bits-and-blooms/bitset.BenchmarkLemireIterateManyb
     1.57s 16.90% 81.27%         3s 32.29%  github.com/bits-and-blooms/bitset.(*BitSet).Set
     1.38s 14.85% 96.12%      1.43s 15.39%  github.com/bits-and-blooms/bitset.(*BitSet).extendSetMaybe  

**After pprofinline**

bitset) go test -bench=LemireIterateManyb -cpu 1 -parallel 1 -benchtime=3x -gcflags="-m -fpprofprofileuse LemireIterate.pprof -pprofinline" -cpuprofile LemireIterate_after.pprof 2> after_log.txt
goos: darwin
goarch: amd64
pkg: github.com/bits-and-blooms/bitset
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkLemireIterateManyb 	       3	1311964698 ns/op
PASS
ok  	github.com/bits-and-blooms/bitset	8.782s
(bitset) go tool pprof LemireIterate_after.pprof
Type: cpu
Time: Dec 6, 2021 at 9:46am (PST)
Duration: 8.02s, Total samples = 6.97s (86.95%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top20
Showing nodes accounting for 6.93s, 99.43% of 6.97s total
Dropped 9 nodes (cum <= 0.03s)
      flat  flat%   sum%        cum   cum%
     3.38s 48.49% 48.49%      3.40s 48.78%  github.com/bits-and-blooms/bitset.(*BitSet).NextSetMany
     3.31s 47.49% 95.98%      6.81s 97.70%  github.com/bits-and-blooms/bitset.BenchmarkLemireIterateManyb

**after_log.txt content**

PprofInline: [hot] inline (*BitSet).extendSetMaybe in (*BitSet).Set with weight 5.20
PprofInline: [hot] inline (*BitSet).Set in BenchmarkLemireIterateManyb with weight 10.92
PprofInline: [hot] inline (*BitSet).NextSetMany in BenchmarkLemireIterateManyb with weight 12.34

Please let me know if they help.

**Comment by lni** [comment]
@rajbarik Thanks for the help above, sadly I still couldn't reproduce the performance improvements. 

Based on your suggestion above, I changed benchtime to 1000x and repeated the test - 

```
lni$ /Users/lni/src/go/bin/go test -bench=LemireIterateManyb -cpu 1 -parallel 1 -benchtime=1000x -gcflags="-m" -cpuprofile LemireIterate.pprof 2> before_log.txt
goos: darwin
goarch: amd64
pkg: github.com/bits-and-blooms/bitset
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
BenchmarkLemireIterateManyb 	    1000	   2275634 ns/op
PASS
ok  	github.com/bits-and-blooms/bitset	3.114s

lni$ go tool pprof LemireIterate.pprof
Type: cpu
Time: Dec 8, 2021 at 11:07pm (CST)
Duration: 2.52s, Total samples = 2.09s (82.86%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top30
Showing nodes accounting for 2.07s, 99.04% of 2.09s total
Dropped 14 nodes (cum <= 0.01s)
      flat  flat%   sum%        cum   cum%
     1.61s 77.03% 77.03%      1.63s 77.99%  github.com/bits-and-blooms/bitset.(*BitSet).NextSetMany
     0.41s 19.62% 96.65%      2.05s 98.09%  github.com/bits-and-blooms/bitset.BenchmarkLemireIterateManyb
     0.03s  1.44% 98.09%      0.03s  1.44%  runtime.asyncPreempt
     0.02s  0.96% 99.04%      0.02s  0.96%  runtime.kevent
         0     0% 99.04%      0.02s  0.96%  runtime.netpoll
         0     0% 99.04%      0.02s  0.96%  runtime.startTheWorldWithSema
         0     0% 99.04%      0.03s  1.44%  runtime.systemstack
         0     0% 99.04%      2.05s 98.09%  testing.(*B).launch
         0     0% 99.04%      2.05s 98.09%  testing.(*B).runN

lni$ /Users/lni/src/go/bin/go test -bench=LemireIterateManyb -cpu 1 -parallel 1 -benchtime=1000x -gcflags="-m -fpprofprofileuse LemireIterate.pprof -pprofinline" -cpuprofile LemireIterate_after.pprof 2> after_log.txt
goos: darwin
goarch: amd64
pkg: github.com/bits-and-blooms/bitset
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
BenchmarkLemireIterateManyb 	    1000	   2290498 ns/op
PASS
ok  	github.com/bits-and-blooms/bitset	2.599s

lni$ go tool pprof LemireIterate_after.pprof
Type: cpu
Time: Dec 8, 2021 at 11:08pm (CST)
Duration: 2.49s, Total samples = 2.05s (82.47%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top40
Showing nodes accounting for 2.03s, 99.02% of 2.05s total
Dropped 9 nodes (cum <= 0.01s)
      flat  flat%   sum%        cum   cum%
     1.48s 72.20% 72.20%      1.52s 74.15%  github.com/bits-and-blooms/bitset.(*BitSet).NextSetMany
     0.48s 23.41% 95.61%      2.04s 99.51%  github.com/bits-and-blooms/bitset.BenchmarkLemireIterateManyb
     0.05s  2.44% 98.05%      0.05s  2.44%  runtime.asyncPreempt
     0.02s  0.98% 99.02%      0.02s  0.98%  runtime.memclrNoHeapPointers
         0     0% 99.02%      0.02s  0.98%  github.com/bits-and-blooms/bitset.(*BitSet).Set
         0     0% 99.02%      0.02s  0.98%  github.com/bits-and-blooms/bitset.(*BitSet).extendSetMaybe
         0     0% 99.02%      0.03s  1.46%  runtime.makeslice
         0     0% 99.02%      0.03s  1.46%  runtime.mallocgc
         0     0% 99.02%      0.02s  0.98%  runtime.memclrNoHeapPointersChunked
         0     0% 99.02%      2.02s 98.54%  testing.(*B).launch
         0     0% 99.02%      0.02s  0.98%  testing.(*B).run1.func1
         0     0% 99.02%      2.04s 99.51%  testing.(*B).runN

lni$ cat after_log.txt | grep NextSetMany
PprofInline: can't determine the hotness, but inline trailingZeroes64 in (*BitSet).NextSetMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkLemireIterateManyb
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkLemireIterateManyb
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannLowDensityIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannLowDensityIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannMidDensityIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannMidDensityIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannMidStrongDensityIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannMidStrongDensityIterateMany

lni$ cat after_log.txt | grep "\[hot\]"
PprofInline: [hot] inline (*BitSet).extendSetMaybe in (*BitSet).Set with weight 3.57
PprofInline: [hot] inline strconv.FormatInt in (*BitSet).String with weight 3.57
PprofInline: [hot] inline (*BitSet).Set in BenchmarkLemireIterateManyb with weight 7.14
PprofInline: [hot] inline (*BitSet).String in TestStringLong with weight 3.57
PprofInline: [hot] inline New in TestBitSetHuge with weight 3.57
```

Please let me know if there anything else I should try. Thanks!

**Comment by zhouguangyuan0718** [comment]
I‘m trying. ButI couldn't reproduce the performance improvements, too.
The result is a little different with lni.
<pre>
$ go test -bench=LemireIterateManyb -cpu 1 -parallel 1 -benchtime=1000x -gcflags="-m" -cpuprofile LemireIterate.pprof 2> before_log.txt
goos: linux
goarch: amd64
pkg: github.com/bits-and-blooms/bitset
cpu: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
BenchmarkLemireIterateManyb         1000           2482742 ns/op
PASS
ok      github.com/bits-and-blooms/bitset       2.754s
$ go tool pprof LemireIterate.pprof
File: bitset.test
Type: cpu
Time: Dec 10, 2021 at 1:18am (CST)
Duration: 2.72s, Total samples = 2.60s (95.47%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof)  top30
Showing nodes accounting for 2.58s, 99.23% of 2.60s total
Dropped 12 nodes (cum <= 0.01s)
      flat  flat%   sum%        cum   cum%
     1.95s 75.00% 75.00%      2.07s 79.62%  github.com/bits-and-blooms/bitset.(*BitSet).NextSetMany
     0.42s 16.15% 91.15%      2.50s 96.15%  github.com/bits-and-blooms/bitset.BenchmarkLemireIterateManyb
     0.12s  4.62% 95.77%      0.12s  4.62%  github.com/bits-and-blooms/bitset.trailingZeroes64 (inline)
     0.09s  3.46% 99.23%      0.09s  3.46%  runtime.memclrNoHeapPointers
         0     0% 99.23%      0.09s  3.46%  github.com/bits-and-blooms/bitset.New
         0     0% 99.23%      0.09s  3.46%  github.com/bits-and-blooms/bitset.TestBitSetHuge
         0     0% 99.23%      0.02s  0.77%  runtime.(*mcache).allocLarge
         0     0% 99.23%      0.10s  3.85%  runtime.makeslice
         0     0% 99.23%      0.10s  3.85%  runtime.mallocgc
         0     0% 99.23%      0.08s  3.08%  runtime.memclrNoHeapPointersChunked
         0     0% 99.23%      2.49s 95.77%  testing.(*B).launch
         0     0% 99.23%      2.50s 96.15%  testing.(*B).runN
         0     0% 99.23%      0.10s  3.85%  testing.tRunner
(pprof) q
go test -bench=LemireIterateManyb -cpu 1 -parallel 1 -benchtime=1000x -gcflags="-m -fpprofprofileuse LemireIterate.pprof -pprofinline" -cpuprofile LemireIterate_after.pprof 2> after_log.txt
goos: linux
goarch: amd64
pkg: github.com/bits-and-blooms/bitset
cpu: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
BenchmarkLemireIterateManyb         1000           2504440 ns/op
PASS
ok      github.com/bits-and-blooms/bitset       2.852s
$  go tool pprof LemireIterate.pprof
File: bitset.test
Type: cpu
Time: Dec 10, 2021 at 1:18am (CST)
Duration: 2.72s, Total samples = 2.60s (95.47%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top30
Showing nodes accounting for 2.58s, 99.23% of 2.60s total
Dropped 12 nodes (cum <= 0.01s)
      flat  flat%   sum%        cum   cum%
     1.95s 75.00% 75.00%      2.07s 79.62%  github.com/bits-and-blooms/bitset.(*BitSet).NextSetMany
     0.42s 16.15% 91.15%      2.50s 96.15%  github.com/bits-and-blooms/bitset.BenchmarkLemireIterateManyb
     0.12s  4.62% 95.77%      0.12s  4.62%  github.com/bits-and-blooms/bitset.trailingZeroes64 (inline)
     0.09s  3.46% 99.23%      0.09s  3.46%  runtime.memclrNoHeapPointers
         0     0% 99.23%      0.09s  3.46%  github.com/bits-and-blooms/bitset.New
         0     0% 99.23%      0.09s  3.46%  github.com/bits-and-blooms/bitset.TestBitSetHuge
         0     0% 99.23%      0.02s  0.77%  runtime.(*mcache).allocLarge
         0     0% 99.23%      0.10s  3.85%  runtime.makeslice
         0     0% 99.23%      0.10s  3.85%  runtime.mallocgc
         0     0% 99.23%      0.08s  3.08%  runtime.memclrNoHeapPointersChunked
         0     0% 99.23%      2.49s 95.77%  testing.(*B).launch
         0     0% 99.23%      2.50s 96.15%  testing.(*B).runN
         0     0% 99.23%      0.10s  3.85%  testing.tRunner
(pprof) q
$  cat after_log.txt | grep NextSetMany
PprofInline Caller=github.com/bits-and-blooms/bitset.BenchmarkLemireIterateManyb Callee=github.com/bits-and-blooms/bitset.(*BitSet).NextSetMany Weight=2200000000
PprofInline: can't determine the hotness, but inline trailingZeroes64 in (*BitSet).NextSetMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkLemireIterateManyb
PprofInline: [hot] inline (*BitSet).NextSetMany in BenchmarkLemireIterateManyb with weight 28.03
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannLowDensityIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannLowDensityIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannMidDensityIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannMidDensityIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannMidStrongDensityIterateMany
PprofInline: can't determine the hotness, but inline (*BitSet).NextSetMany in BenchmarkFlorianUekermannMidStrongDensityIterateMany
$   cat after_log.txt | grep "\[hot\]"
PprofInline: [hot] inline (*BitSet).NextSetMany in BenchmarkLemireIterateManyb with weight 28.03
</pre>

@rajbarik
I want to know what else I should try, too. 
Thank you.

**Comment by zhouguangyuan0718** [comment]
For `github.com/bits-and-blooms/bitset`. I run benchmark for all the testcase. And add a test flag -count=100.

<pre>
$ go test -bench=. -cpu 1 -parallel 1 -benchtime=3x -count=100 -cpuprofile all.pprof > before.txt
$ go test -bench=. -cpu 1 -parallel 1 -benchtime=3x -count=100 -gcflags="-fpprofprofileuse all.pprof -specialize -pprofinline" >after.txt
$ benchstat ./before.txt after.txt
name                                             old time/op   new time/op  delta
Set                                               91.4ns ±52%  55.8ns ±40%  -38.89%  (p=0.000 n=95+95)
GetTest                                           81.6ns ±38%  62.2ns ±39%  -23.72%  (p=0.000 n=98+98)
SetExpand                                        1.02µs ±153%  1.95µs ±15%  +90.68%  (p=0.000 n=98+91)
Count                                             1.16µs ± 1%  1.16µs ± 2%     ~     (p=0.777 n=90+96)
Iterate                                           13.1µs ± 0%  13.1µs ± 0%   -0.06%  (p=0.003 n=92+86)
SparseIterate                                     14.5µs ±30%  13.6µs ± 1%   -6.49%  (p=0.000 n=99+84)
LemireCreate                                      14.9ms ± 6%  15.7ms ± 4%   +5.62%  (p=0.000 n=95+97)
LemireCount                                       1.54ms ± 9%  1.51ms ± 7%   -1.85%  (p=0.000 n=98+100)
LemireIterate                                     4.20ms ± 1%  4.16ms ± 1%   -1.11%  (p=0.000 n=88+99)
LemireIterateb                                    4.22ms ± 1%  4.17ms ± 1%   -1.11%  (p=0.000 n=92+100)
LemireIterateManyb                                2.94ms ± 5%  2.98ms ± 2%   +1.05%  (p=0.000 n=94+96)
FlorianUekermannIterateMany                        724ns ±10%   679ns ± 9%   -6.24%  (p=0.000 n=89+98)
FlorianUekermannIterateManyReg                    1.46µs ± 4%  1.40µs ± 4%   -4.45%  (p=0.000 n=100+97)
FlorianUekermannIterateManyComp                    788ns ±11%   926ns ± 9%  +17.47%  (p=0.000 n=97+100)
FlorianUekermannLowDensityIterateMany             1.16ms ± 1%  1.19ms ± 1%   +2.54%  (p=0.000 n=91+97)
FlorianUekermannLowDensityIterateManyReg          1.14ms ± 1%  1.13ms ± 1%   -0.74%  (p=0.000 n=90+98)
FlorianUekermannLowDensityIterateManyComp         1.34ms ± 1%  1.35ms ± 1%   +0.54%  (p=0.000 n=94+93)
FlorianUekermannMidDensityIterateMany             12.7ms ± 1%  12.9ms ± 1%   +1.79%  (p=0.000 n=89+100)
FlorianUekermannMidDensityIterateManyReg          19.0ms ± 1%  18.9ms ± 0%   -0.21%  (p=0.000 n=94+97)
FlorianUekermannMidDensityIterateManyComp         13.8ms ± 1%  13.7ms ± 0%   -0.45%  (p=0.000 n=92+99)
FlorianUekermannMidStrongDensityIterateMany       38.3ms ± 1%  38.4ms ± 0%   +0.36%  (p=0.000 n=96+98)
FlorianUekermannMidStrongDensityIterateManyReg    71.1ms ± 0%  74.8ms ± 0%   +5.26%  (p=0.000 n=95+97)
FlorianUekermannMidStrongDensityIterateManyComp   46.3ms ± 2%  46.4ms ± 1%   +0.21%  (p=0.000 n=99+92)

</pre>

For some testcase, it improved. But some  testcases got a worse result.

**Comment by lni** [comment]
@rajbarik

The bitset version I am using is 5a829244ffd64e4120015ce1bf8285b0b6168d55 which is the latest in its default branch. Any chance you can quickly check which version of bitset are you running? 

Also wondering is there any other tests/benchmarks that are known to benefit from the PGO patch? Will be happy to try those as well. Thanks!

**Comment by rajbarik** [comment]
@lni I will take a look at this next week and update you. Please stay tuned.

**Comment by lni** [comment]
@rajbarik Thanks a lot for looking into this. 

We are building a new [HTAP database in Go called MatrixOne](https://github.com/matrixorigin/matrixone). Our profiling results indicate that some hot functions are not inlined as expected when using the official compiler, in our experiments, we manually batched their invocations and immediately observed noticeable performance improvements. Profiling guided inlining as described in your PGO patch sounds like a pretty cool potential solution to the problem we face. 

Looking forward to your update, will be very happy to further cooperate on this exciting PGO feature!

**Comment by rajbarik** [comment]
@lni @zhouguangyuan0718 My apologies, I had modified the BenchmarkLemireIterateManyb function in bitset_benchmark_test.go to increase its execution time by setting everybit in bitset. I had modified one line; please see the commented line below.

`// go test -bench=BenchmarkLemireIterateManyb
// see http://lemire.me/blog/2016/09/22/swift-versus-java-the-bitset-performance-test/
func BenchmarkLemireIterateManyb(b *testing.B) {
  bitmap := New(1000000000)
  for v := uint(0); v <= 1000000000; v += 1 { // This is the only change to the code which replaces 100 with 1
    bitmap.Set(v)
  }
  buffer := make([]uint, 256)
  b.ResetTimer()
  sum := uint(0)
  for i := 0; i < b.N; i++ {
    j := uint(0)
    j, buffer = bitmap.NextSetMany(j, buffer)
    for ; len(buffer) > 0; j, buffer = bitmap.NextSetMany(j, buffer) {
      for k := range buffer {
        sum += buffer[k]
      }
      j++
    }
  }

  if sum == 0 { // added just to fool ineffassign
    return
  }
}`

With this change, here is what I see on my Intel Mac laptop:
`
+ go test -bench=BenchmarkLemireIterateManyb -cpu 1 -parallel 1 -count 5 -benchtime=3x -cpuprofile LemireIterate.pprof > before.txt
+ go test -bench=BenchmarkLemireIterateManyb -cpu 1 -parallel 1 -count 5 -benchtime=3x '-gcflags=-fpprofprofileuse LemireIterate.pprof -pprofinline' -cpuprofile LemireIterate_after.pprof > after.txt
+ benchstat before.txt after.txt
name                old time/op  new time/op  delta
LemireIterateManyb   1.46s ± 4%   1.31s ± 2%  -9.87%  (p=0.008 n=5+5)
`


Another benchmark where I see ~25% improvement is [Tally](https://github.com/uber-go/tally). Steps below:
`
+ go test -bench=ScopeReporting/size1000000 -cpu 1 -parallel 1 -count 5 -benchtime=1x -cpuprofile scope.pprof > before.txt
+ go test -bench=ScopeReporting/size1000000 -cpu 1 -parallel 1 -count 5 -benchtime=1x '-gcflags=-fpprofprofileuse scope.pprof -pprofinline' -cpuprofile scope_after.pprof > after.txt
+ benchstat before.txt after.txt
name                        old time/op  new time/op  delta
ScopeReporting/size1000000  2.75ms ± 5%  2.04ms ± 5%  -25.75%  (p=0.008 n=5+5)
`

Please let me know if you are able to reproduce bitset and tally now.

@lni  Regarding MatrixOne, I am able to run the unit-tests. Is there any particular unit-test I should dig deeper? Are there integration/component tests?

**Comment by mvdan** [comment]
It seems like PGO is being talked about internally again (https://github.com/golang/go/issues/43930#issuecomment-1209759830). Is there a chance that we could have more details about the current plan and design? For those of us interested in the Go compiler while outside Google, it's being particularly hard to follow what is happening, let alone contribute in any meaningful way :)

I understand that it may still be early for a formal proposal, but perhaps there could be a regular GitHub issue thread where we can subscribe for detailed updates every few weeks. Ideally the discussions about PGO would simply happen in a public place, removing the need for that back and forth between private and public communication.

**Comment by mvdan** [comment]
It's only been two days, but my [last comment](https://github.com/golang/go/pull/49688#issuecomment-1222093959) is already buried by the commit push activity :) I had to click on "see more" three times to find it again. Can we please at least have one tracking issue to talk about PGO?

**Comment by josharian** [comment]
https://github.com/golang/go/issues/28262

**Comment by mvdan** [comment]
Oh, thanks Josh. It has been some time and I even forgot I already asked this last year :)

**Comment by prattmic** [comment]
FYI, this has moved to https://go.dev/cl/429863, and there is a proposal at #55022.
